File: apps/askbible-ios/AskBible/Audio/MusicPlayer.swift
import AVFoundation
import Combine
import MediaPlayer

/// 音乐页播放器。曲库是 MusicCatalog（内置 5 首 + R2 点播），队列 = 当前专辑的全部曲目。
///
/// 与整章朗读的 ChapterAudioPlayer 是两个独立播放器（RN 版是一个 shell 播放器切模式）；
/// 互斥靠 `onWillPlay` 由 RootView 接线，锁屏命令靠 RemoteControlHub 路由。
/// 专辑规则（循环 / 音量 / 睡眠定时联动 / 起播曲）见 MusicAlbumRules，与 RN 的 useMusicHomeAlbum 一致。
@MainActor
final class MusicPlayer: ObservableObject {
    @Published private(set) var album: String
    /// 当前曲在 MusicCatalog.tracks 里的下标
    @Published private(set) var trackIndex: Int
    @Published private(set) var isPlaying = false
    @Published private(set) var isLoading = false
    @Published private(set) var currentTime: Double = 0
    @Published private(set) var duration: Double = 0
    @Published private(set) var errorMessage: String?
    @Published private(set) var repeatMode: MusicRepeatMode
    @Published private(set) var interrupted = false
    @Published private(set) var sleepDeadline: Date?
    /// 用户选的睡眠档位（0 = 未设），切专辑的联动规则要看它
    private var sleepMinutes = 0
    private var sleepTimer: Timer?

    /// 开播前先让别的播放器（整章朗读）停下
    var onWillPlay: (() -> Void)?

    private var player: AVPlayer?
    private var timeObserver: Any?
    private var endObserver: NSObjectProtocol?
    private var statusObserver: NSKeyValueObservation?
    private var loadedId: String?
    private var wantsPlayback = false
    private var gain: Float

    init() {
        let a = MusicCatalog.defaultAlbum
        album = a
        trackIndex = MusicAlbumRules.startIndex(MusicCatalog.tracks, album: a, current: -1) ?? 0
        repeatMode = MusicAlbumRules.defaultRepeatMode(a) ?? .all
        gain = MusicAlbumRules.defaultGain(a)
        observeInterruptions()
    }

    deinit {
        if let timeObserver { player?.removeTimeObserver(timeObserver) }
        if let endObserver { NotificationCenter.default.removeObserver(endObserver) }
    }

    // MARK: 队列

    /// 当前专辑的全部曲目（全曲库下标）
    var queue: [Int] {
        MusicCatalog.tracks.indices.filter { MusicCatalog.tracks[$0].album == album }
    }

    var track: MusicTrack? {
        MusicCatalog.tracks.indices.contains(trackIndex) ? MusicCatalog.tracks[trackIndex] : nil
    }

    /// 队列里的邻居，首尾相接；队列不足两首返回 nil
    private func neighbor(_ step: Int) -> Int? {
        let q = queue
        guard q.count > 1, let pos = q.firstIndex(of: trackIndex) else { return nil }
        return q[(pos + step + q.count) % q.count]
    }

    var previousTrack: MusicTrack? { neighbor(-1).map { MusicCatalog.tracks[$0] } }
    var nextTrack: MusicTrack? { neighbor(1).map { MusicCatalog.tracks[$0] } }

    /// 还没探到真实时长前先用曲库元数据，进度条右侧不闪「—:—」
    var displayDuration: Double {
        duration > 0 ? duration : Double(track?.durationSec ?? 0)
    }

    var progress: Double {
        let d = displayDuration
        return d > 0 ? min(1, max(0, currentTime / d)) : 0
    }

    // MARK: 专辑

    /// 切专辑（RN useMusicHomeAlbum.selectAlbum）：换循环模式、联动睡眠定时、换音量、
    /// 跳到该专辑起播曲；正在播就接着播，不然停在暂停态。
    func selectAlbum(_ raw: String) {
        let next = MusicAlbumRules.normalize(raw)
        guard next != album else { return }
        album = next
        if let mode = MusicAlbumRules.defaultRepeatMode(next) { repeatMode = mode }
        if let mins = MusicAlbumRules.sleepTimerOnSwitch(to: next, currentMinutes: sleepMinutes) {
            setSleepTimer(minutes: mins == 0 ? nil : mins)
        }
        gain = MusicAlbumRules.defaultGain(next)
        player?.volume = gain
        if let idx = MusicAlbumRules.startIndex(MusicCatalog.tracks, album: next, current: trackIndex) {
            play(index: idx, autoPlay: isPlaying || wantsPlayback)
        }
    }

    // MARK: 传输

    /// 音乐页播放键（RN resolveMusicPageToggleAction）：
    /// 正在播且当前曲就在选中专辑 → 暂停；否则在选中专辑里起播（当前曲在专辑内就接着它）。
    func toggle() {
        if isPlaying, track?.album == album { pause(); return }
        if let t = track, t.album == album {
            if loadedId == t.id { resume() } else { play(index: trackIndex, autoPlay: true) }
            return
        }
        if let idx = MusicAlbumRules.startIndex(MusicCatalog.tracks, album: album, current: trackIndex) {
            play(index: idx, autoPlay: true)
        }
    }

    func next() {
        guard let n = neighbor(1) else { seek(to: 0); return }
        play(index: n, autoPlay: isPlaying || wantsPlayback)
    }

    func previous() {
        // 与常见播放器一致：播过 3 秒先回本曲开头
        if currentTime > 3 { seek(to: 0); return }
        guard let p = neighbor(-1) else { seek(to: 0); return }
        play(index: p, autoPlay: isPlaying || wantsPlayback)
    }

    func select(index: Int) {
        guard MusicCatalog.tracks.indices.contains(index) else { return }
        play(index: index, autoPlay: true)
    }

    func toggleRepeatOne() { repeatMode = repeatMode == .one ? .off : .one }
    func toggleRepeatAll() { repeatMode = repeatMode == .all ? .off : .all }

    /// 装载并（可选）起播某一首
    func play(index: Int, autoPlay: Bool) {
        guard MusicCatalog.tracks.indices.contains(index) else { return }
        let t = MusicCatalog.tracks[index]
        trackIndex = index
        if loadedId != t.id { load(t) }
        if autoPlay { resume() }
    }

    private func load(_ t: MusicTrack) {
        teardown()
        loadedId = t.id
        errorMessage = nil
        currentTime = 0
        duration = 0
        guard let url = MusicAudioSource.url(for: t) else {
            errorMessage = "这首曲子没有可用音源"
            return
        }
        isLoading = !url.isFileURL
        let asset = AVURLAsset(url: url, options: [
            "AVURLAssetHTTPHeaderFieldsKey": ["User-Agent": MusicAudioSource.userAgent]
        ])
        let item = AVPlayerItem(asset: asset)
        let avPlayer = AVPlayer(playerItem: item)
        avPlayer.automaticallyWaitsToMinimizeStalling = true
        avPlayer.volume = gain
        player = avPlayer

        statusObserver = item.observe(\.status, options: [.new]) { [weak self] item, _ in
            Task { @MainActor in
                guard let self else { return }
                switch item.status {
                case .readyToPlay:
                    self.isLoading = false
                    let d = item.duration.seconds
                    self.duration = d.isFinite && d > 0 ? d : 0
                    self.updateNowPlaying()
                case .failed:
                    self.isLoading = false
                    self.errorMessage = item.error?.localizedDescription ?? "音乐加载失败"
                    self.isPlaying = false
                default:
                    break
                }
            }
        }

        timeObserver = avPlayer.addPeriodicTimeObserver(
            forInterval: CMTime(seconds: 0.25, preferredTimescale: 600), queue: .main
        ) { [weak self] time in
            Task { @MainActor in
                guard let self else { return }
                self.currentTime = time.seconds.isFinite ? time.seconds : 0
                if self.duration == 0, let d = self.player?.currentItem?.duration.seconds, d.isFinite, d > 0 {
                    self.duration = d
                }
            }
        }

        endObserver = NotificationCenter.default.addObserver(
            forName: .AVPlayerItemDidPlayToEndTime, object: item, queue: .main
        ) { [weak self] _ in
            Task { @MainActor in self?.handleEnded() }
        }
    }

    /// 播完一首：单曲循环回头；整专辑循环顺延（首尾相接）；不循环则顺延到队尾停下
    private func handleEnded() {
        switch repeatMode {
        case .one:
            seek(to: 0)
            resume()
        case .all:
            if let n = neighbor(1) { play(index: n, autoPlay: true) } else { seek(to: 0); resume() }
        case .off:
            let q = queue
            if let pos = q.firstIndex(of: trackIndex), pos + 1 < q.count {
                play(index: q[pos + 1], autoPlay: true)
            } else {
                isPlaying = false
                wantsPlayback = false
                currentTime = duration
                updateNowPlaying()
            }
        }
    }

    /// 用户点播放（或打断结束自动续播）：不看 `interrupted` —— 通话 / Siri / 别的 App 抢声道后 iOS 常常不发「打断结束」，
    /// 之前这里 guard 一下就把播放键永久点死了（Josh 真机 2026-09-10「点播放播放不了」）。RN 也只挡后台自动续播，不挡用户点。
    func resume() {
        guard player != nil else { return }
        wantsPlayback = true
        interrupted = false
        onWillPlay?()
        try? AVAudioSession.sharedInstance().setCategory(.playback, mode: .default)
        try? AVAudioSession.sharedInstance().setActive(true)
        RemoteControlHub.shared.claim(self)
        player?.volume = gain
        player?.play()
        isPlaying = true
        updateNowPlaying()
    }

    func pause() {
        wantsPlayback = false
        player?.pause()
        isPlaying = false
        updateNowPlaying()
    }

    func seek(ratio: Double) {
        let d = displayDuration
        guard d > 0 else { return }
        seek(to: min(1, max(0, ratio)) * d)
    }

    func seek(to seconds: Double) {
        let d = displayDuration
        let clamped = d > 0 ? min(max(0, seconds), d) : max(0, seconds)
        player?.seek(to: CMTime(seconds: clamped, preferredTimescale: 600))
        currentTime = clamped
        updateNowPlaying()
    }

    // MARK: 睡眠定时（与 ChapterAudioPlayer 同一套：到期只暂停）

    func setSleepTimer(minutes: Int?) {
        sleepTimer?.invalidate()
        sleepTimer = nil
        guard let minutes, minutes > 0 else { sleepMinutes = 0; sleepDeadline = nil; return }
        sleepMinutes = minutes
        let deadline = Date().addingTimeInterval(TimeInterval(minutes * 60))
        sleepDeadline = deadline
        sleepTimer = Timer.scheduledTimer(withTimeInterval: deadline.timeIntervalSinceNow, repeats: false) { [weak self] _ in
            Task { @MainActor in
                guard let self else { return }
                self.sleepDeadline = nil
                self.sleepMinutes = 0
                self.pause()
            }
        }
    }

    var sleepRemainingLabel: String? {
        guard let d = sleepDeadline else { return nil }
        let s = max(0, Int(d.timeIntervalSinceNow.rounded()))
        return String(format: "%d:%02d", s / 60, s % 60)
    }

    // MARK: 系统

    private func observeInterruptions() {
        NotificationCenter.default.addObserver(
            forName: AVAudioSession.interruptionNotification,
            object: AVAudioSession.sharedInstance(), queue: .main
        ) { [weak self] note in
            Task { @MainActor in
                guard let self,
                      let raw = note.userInfo?[AVAudioSessionInterruptionTypeKey] as? UInt,
                      let type = AVAudioSession.InterruptionType(rawValue: raw) else { return }
                switch type {
                case .began:
                    self.interrupted = true
                    if self.isPlaying {
                        self.player?.pause()
                        self.isPlaying = false
                        self.updateNowPlaying()
                    }
                case .ended:
                    self.interrupted = false
                    let opts = (note.userInfo?[AVAudioSessionInterruptionOptionKey] as? UInt)
                        .map(AVAudioSession.InterruptionOptions.init(rawValue:)) ?? []
                    if opts.contains(.shouldResume), self.wantsPlayback { self.resume() }
                @unknown default:
                    break
                }
            }
        }
        NotificationCenter.default.addObserver(
            forName: AVAudioSession.routeChangeNotification,
            object: AVAudioSession.sharedInstance(), queue: .main
        ) { [weak self] note in
            Task { @MainActor in
                guard let self,
                      let raw = note.userInfo?[AVAudioSessionRouteChangeReasonKey] as? UInt,
                      AVAudioSession.RouteChangeReason(rawValue: raw) == .oldDeviceUnavailable
                else { return }
                self.pause()
            }
        }
    }

    private func updateNowPlaying() {
        guard let t = track else { return }
        var info: [String: Any] = [
            MPMediaItemPropertyTitle: t.title,
            MPMediaItemPropertyArtist: t.artist.isEmpty ? "AskBible" : t.artist,
            MPMediaItemPropertyAlbumTitle: t.album,
            MPNowPlayingInfoPropertyElapsedPlaybackTime: currentTime,
            MPNowPlayingInfoPropertyPlaybackRate: isPlaying ? 1.0 : 0.0,
        ]
        if displayDuration > 0 { info[MPMediaItemPropertyPlaybackDuration] = displayDuration }
        MPNowPlayingInfoCenter.default().nowPlayingInfo = info
    }

    private func teardown() {
        if let timeObserver { player?.removeTimeObserver(timeObserver) }
        if let endObserver { NotificationCenter.default.removeObserver(endObserver) }
        statusObserver?.invalidate()
        timeObserver = nil
        endObserver = nil
        statusObserver = nil
        player?.pause()
        player = nil
        loadedId = nil
        isPlaying = false
    }
}

extension MusicPlayer: RemotePlayable {
    func remotePlay() { resume() }
    func remotePause() { pause() }
    func remoteToggle() { isPlaying ? pause() : resume() }
    func remoteNext() { next() }
    func remotePrevious() { previous() }
    func remoteSeek(to seconds: Double) { seek(to: seconds) }
}

File: apps/askbible-android/app/src/main/java/me/askbible/native_/audio/MusicPlayer.kt
package me.askbible.native_.audio

import me.askbible.native_.data.SiteCopy
import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.media3.common.AudioAttributes
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.MediaMetadata
import androidx.media3.common.Player
import androidx.media3.datasource.DefaultDataSource
import androidx.media3.datasource.DefaultHttpDataSource
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.exoplayer.source.DefaultMediaSourceFactory
import androidx.media3.session.MediaSession
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import me.askbible.native_.data.MusicAlbumRules
import me.askbible.native_.data.MusicAudioSource
import me.askbible.native_.data.MusicCatalog
import me.askbible.native_.data.MusicRepeatMode
import me.askbible.native_.data.MusicTrack

/**
 * 音乐页播放器（ExoPlayer）。与 iOS 的 MusicPlayer 对等：曲库 MusicCatalog（内置 5 首 + R2 点播），
 * 队列 = 当前专辑全部曲目；专辑规则见 MusicAlbumRules（与 RN useMusicHomeAlbum 一致）。
 *
 * 与整章朗读的 ChapterAudioPlayer 是两个独立播放器，互斥靠 onWillPlay 由 RootScreen 接线；
 * 各自一个 MediaSession（id 不同），系统媒体控制显示最近活跃的那个。
 */
class MusicPlayer(context: Context, private val scope: CoroutineScope) {
    private val appContext = context.applicationContext

    var album by mutableStateOf(MusicCatalog.DEFAULT_ALBUM); private set
    /** 当前曲在 MusicCatalog.tracks 里的下标 */
    var trackIndex by mutableStateOf(0); private set
    var isPlaying by mutableStateOf(false); private set
    var isLoading by mutableStateOf(false); private set
    var currentTime by mutableStateOf(0.0); private set
    var duration by mutableStateOf(0.0); private set
    var errorMessage by mutableStateOf<String?>(null); private set
    var repeatMode by mutableStateOf(MusicRepeatMode.ALL); private set
    var sleepDeadlineMs by mutableStateOf<Long?>(null); private set
    /** 用户选的睡眠档位（0 = 未设），切专辑的联动规则要看它 */
    private var sleepMinutes = 0
    private var sleepJob: Job? = null

    /** 开播前先让别的播放器（整章朗读）停下 */
    var onWillPlay: (() -> Unit)? = null

    private var loadedId: String? = null
    private var wantsPlayback = false
    /** 被系统「永久」夺走焦点而停的（还想播）：外部声音停了 / 回前台就续播 */
    private var focusLost = false
    private var gain = 1f
    private var ticker: Job? = null

    private val player: ExoPlayer = ExoPlayer.Builder(context)
        .setMediaSourceFactory(
            DefaultMediaSourceFactory(
                // DefaultDataSource 同时认 asset:///（内置曲）与 http(s)（R2 / Hymn Commons）
                DefaultDataSource.Factory(
                    context,
                    DefaultHttpDataSource.Factory()
                        .setUserAgent(MusicAudioSource.USER_AGENT)
                        .setAllowCrossProtocolRedirects(true)
                )
            )
        )
        .build()

    private val mediaSession = MediaSession.Builder(context, player).setId("music").build()
        .also { PlaybackSessions.register(it) { wantsPlayback } }

    init {
        val a = MusicCatalog.DEFAULT_ALBUM
        trackIndex = MusicAlbumRules.startIndex(MusicCatalog.tracks, a, -1) ?: 0
        repeatMode = MusicAlbumRules.defaultRepeatMode(a) ?: MusicRepeatMode.ALL
        gain = MusicAlbumRules.defaultGain(a)
        player.setAudioAttributes(
            AudioAttributes.Builder()
                .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
                .setUsage(C.USAGE_MEDIA)
                .build(),
            true
        )
        player.addListener(object : Player.Listener {
            override fun onIsPlayingChanged(playing: Boolean) {
                this@MusicPlayer.isPlaying = playing
            }

            override fun onPlayWhenReadyChanged(playWhenReady: Boolean, reason: Int) {
                if (playWhenReady) focusLost = false
                else if (reason == Player.PLAY_WHEN_READY_CHANGE_REASON_AUDIO_FOCUS_LOSS && wantsPlayback) focusLost = true
            }

            override fun onPlaybackStateChanged(state: Int) {
                this@MusicPlayer.isLoading = state == Player.STATE_BUFFERING
                if (state == Player.STATE_READY && this@MusicPlayer.duration <= 0) {
                    val d = player.duration
                    if (d > 0) this@MusicPlayer.duration = d / 1000.0
                }
                if (state == Player.STATE_ENDED) onEnded()
            }

            override fun onPlayerError(error: androidx.media3.common.PlaybackException) {
                this@MusicPlayer.errorMessage = error.errorCodeName
                this@MusicPlayer.isPlaying = false
            }
        })
    }

    // ---- 队列 ----

    /** 当前专辑的全部曲目（全曲库下标） */
    val queue: List<Int> get() = MusicCatalog.tracks.indices.filter { MusicCatalog.tracks[it].album == album }

    val track: MusicTrack? get() = MusicCatalog.tracks.getOrNull(trackIndex)

    /** 队列里的邻居，首尾相接；队列不足两首返回 null */
    private fun neighbor(step: Int): Int? {
        val q = queue
        if (q.size < 2) return null
        val pos = q.indexOf(trackIndex)
        if (pos < 0) return null
        return q[((pos + step) % q.size + q.size) % q.size]
    }

    val previousTrack: MusicTrack? get() = neighbor(-1)?.let { MusicCatalog.tracks[it] }
    val nextTrack: MusicTrack? get() = neighbor(1)?.let { MusicCatalog.tracks[it] }

    /** 还没探到真实时长前先用曲库元数据，进度条右侧不闪「—:—」 */
    val displayDuration: Double get() = if (duration > 0) duration else (track?.durationSec ?: 0).toDouble()

    val progress: Double get() {
        val d = displayDuration
        return if (d > 0) (currentTime / d).coerceIn(0.0, 1.0) else 0.0
    }

    // ---- 专辑 ----

    /**
     * 切专辑（RN useMusicHomeAlbum.selectAlbum）：换循环模式、联动睡眠定时、换音量、
     * 跳到该专辑起播曲；正在播就接着播，不然停在暂停态。
     */
    fun selectAlbum(raw: String) {
        val next = MusicAlbumRules.normalize(raw)
        if (next == album) return
        album = next
        MusicAlbumRules.defaultRepeatMode(next)?.let { repeatMode = it }
        MusicAlbumRules.sleepTimerOnSwitch(next, sleepMinutes)?.let { setSleepTimer(if (it == 0) null else it) }
        gain = MusicAlbumRules.defaultGain(next)
        player.volume = gain
        MusicAlbumRules.startIndex(MusicCatalog.tracks, next, trackIndex)?.let { play(it, isPlaying || wantsPlayback) }
    }

    // ---- 传输 ----

    /**
     * 音乐页播放键（RN resolveMusicPageToggleAction）：
     * 正在播且当前曲就在选中专辑 → 暂停；否则在选中专辑里起播（当前曲在专辑内就接着它）。
     */
    fun toggle() {
        val t = track
        if (isPlaying && t?.album == album) { pause(); return }
        if (t != null && t.album == album) {
            if (loadedId == t.id) resume() else play(trackIndex, true)
            return
        }
        MusicAlbumRules.startIndex(MusicCatalog.tracks, album, trackIndex)?.let { play(it, true) }
    }

    fun next() {
        val n = neighbor(1) ?: run { seekTo(0.0); return }
        play(n, isPlaying || wantsPlayback)
    }

    fun previous() {
        // 与常见播放器一致：播过 3 秒先回本曲开头
        if (currentTime > 3) { seekTo(0.0); return }
        val p = neighbor(-1) ?: run { seekTo(0.0); return }
        play(p, isPlaying || wantsPlayback)
    }

    fun select(index: Int) {
        if (index !in MusicCatalog.tracks.indices) return
        play(index, true)
    }

    fun toggleRepeatOne() { repeatMode = if (repeatMode == MusicRepeatMode.ONE) MusicRepeatMode.OFF else MusicRepeatMode.ONE }
    fun toggleRepeatAll() { repeatMode = if (repeatMode == MusicRepeatMode.ALL) MusicRepeatMode.OFF else MusicRepeatMode.ALL }

    /** 装载并（可选）起播某一首 */
    fun play(index: Int, autoPlay: Boolean) {
        val t = MusicCatalog.tracks.getOrNull(index) ?: return
        trackIndex = index
        if (loadedId != t.id) load(t)
        if (autoPlay) resume()
    }

    private fun load(t: MusicTrack) {
        loadedId = t.id
        errorMessage = null
        currentTime = 0.0
        duration = 0.0
        val uri = MusicAudioSource.uri(t)
        if (uri == null) {
            errorMessage = SiteCopy.t("native.trackNoSource")
            player.stop(); player.clearMediaItems()
            return
        }
        player.setMediaItem(
            MediaItem.Builder()
                .setUri(uri)
                .setMediaMetadata(
                    MediaMetadata.Builder()
                        .setTitle(t.localizedTitle)
                        .setArtist(t.artist.ifEmpty { "AskBible" })
                        .setAlbumTitle(t.album)
                        .build())
                .build()
        )
        player.volume = gain
        player.prepare()
        startTicker()
    }

    /** 播完一首：单曲循环回头；整专辑循环顺延（首尾相接）；不循环则顺延到队尾停下 */
    private fun onEnded() {
        when (repeatMode) {
            MusicRepeatMode.ONE -> { seekTo(0.0); resume() }
            MusicRepeatMode.ALL -> { val n = neighbor(1); if (n != null) play(n, true) else { seekTo(0.0); resume() } }
            MusicRepeatMode.OFF -> {
                val q = queue
                val pos = q.indexOf(trackIndex)
                if (pos >= 0 && pos + 1 < q.size) play(q[pos + 1], true)
                else { wantsPlayback = false; currentTime = duration }
            }
        }
    }

    fun resume() {
        if (loadedId == null) { track?.let { load(it) } ?: return }
        wantsPlayback = true
        onWillPlay?.invoke()
        PlaybackSessions.ensureStarted(appContext)
        player.volume = gain
        player.playWhenReady = true
    }

    fun pause() {
        wantsPlayback = false
        focusLost = false
        player.playWhenReady = false
    }

    /** 外部声音停了 / 回到前台：永久失焦停掉的、还想播的，叫回来（RN recoverMusicPlaybackAfterBackground） */
    fun recoverAfterInterruption() {
        if (!focusLost || !wantsPlayback || isPlaying) return
        if (AudioInterruptionMonitor.callLike || AudioInterruptionMonitor.foreignAudioActive) return
        focusLost = false
        resume()
    }

    fun seekRatio(ratio: Double) {
        val d = displayDuration
        if (d <= 0) return
        seekTo(ratio.coerceIn(0.0, 1.0) * d)
    }

    fun seekTo(seconds: Double) {
        val d = displayDuration
        val clamped = if (d > 0) seconds.coerceIn(0.0, d) else maxOf(0.0, seconds)
        player.seekTo((clamped * 1000).toLong())
        currentTime = clamped
    }

    // ---- 睡眠定时（与 ChapterAudioPlayer 同一套：到期只暂停） ----

    fun setSleepTimer(minutes: Int?) {
        sleepJob?.cancel(); sleepJob = null
        if (minutes == null || minutes <= 0) { sleepMinutes = 0; sleepDeadlineMs = null; return }
        sleepMinutes = minutes
        val deadline = System.currentTimeMillis() + minutes * 60_000L
        sleepDeadlineMs = deadline
        sleepJob = scope.launch(Dispatchers.Main) {
            delay(deadline - System.currentTimeMillis())
            sleepDeadlineMs = null
            sleepMinutes = 0
            pause()
        }
    }

    val sleepRemainingLabel: String?
        get() {
            val d = sleepDeadlineMs ?: return null
            val s = ((d - System.currentTimeMillis()) / 1000).coerceAtLeast(0)
            return "%d:%02d".format(s / 60, s % 60)
        }

    /** ExoPlayer 没有 addPeriodicTimeObserver，用协程轮询（间隔与 iOS 一致） */
    private fun startTicker() {
        ticker?.cancel()
        ticker = scope.launch(Dispatchers.Main) {
            while (true) {
                currentTime = player.currentPosition / 1000.0
                if (duration <= 0 && player.duration > 0) duration = player.duration / 1000.0
                delay(250)
            }
        }
    }

    fun release() {
        sleepJob?.cancel()
        ticker?.cancel()
        PlaybackSessions.unregister(mediaSession)
        mediaSession.release()
        player.release()
    }
}
